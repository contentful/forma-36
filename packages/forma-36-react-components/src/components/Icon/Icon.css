@import 'settings/colors';

.Icon {
  display: inline-block;
  fill: var(--color-blue-base);
}

.Icon--tiny {
  height: 16px;
  width: 16px;
}

.Icon--small {
  height: 18px;
  width: 18px;
}

.Icon--medium {
  height: 24px;
  width: 24px;
}

.Icon--large {
  height: 32px;
  width: 32px;
}

.Icon--trimmed {
  width: auto;
}

.Icon--positive {
  fill: var(--color-positive);
}

.Icon--negative {
  fill: var(--color-negative);
}

.Icon--warning {
  fill: var(--color-warning);
}

.Icon--secondary {
  fill: var(--color-text-mid);
}

.Icon--muted {
  fill: var(--color-text-light);
}

.Icon--white {
  fill: var(--color-white);
}
