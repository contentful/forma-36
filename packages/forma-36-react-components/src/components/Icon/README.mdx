Icons are used to indicate actions, statuses and navigation. The `Icon` component can be used in `Button`, `TextLink`, `Note` and other components.

## Table of contents

- [How to use Icon](#how-to-use-Icon)
- [Component variations](#component-variations)
- [Accessibility](#accessibility)

## How to use Icon

- Select an icon that accurately represents the subject
- Pair icons with text
- Position buttons consistently in the interface
- Ensure the behavior of the button is consistent in all use cases
- Consider how the button fits into the context of the screen and reduce complexity where possible

## Component variations

The `Icon` component can be configured in different ways using variations in color, size and icon type:

### Icon colors

```jsx
<React.Fragment>
  <div>
    <Icon color="primary" icon="Edit" /> Primary
  </div>
  <div>
    <Icon color="positive" icon="Edit" /> Positive
  </div>
  <div>
    <Icon color="negative" icon="Edit" /> Negative
  </div>
  <div>
    <Icon color="warning" icon="Edit" /> Warning
  </div>
  <div>
    <Icon color="secondary" icon="Edit" /> Secondary
  </div>
  <div>
    <Icon color="muted" icon="Edit" /> Muted
  </div>
  <div style={{ backgroundColor: '#8091a5' }}>
    <Icon color="white" icon="Edit" /> White
  </div>
</React.Fragment>
```

### Icon sizes

```jsx
<React.Fragment>
  <div>
    <Icon color="primary" size="tiny" icon="Edit" /> Tiny size
  </div>
  <div>
    <Icon color="primary" size="small" icon="Edit" /> Small size
  </div>
  <div>
    <Icon color="primary" size="large" icon="Edit" /> Large size
  </div>
</React.Fragment>
```

### Icon type

```jsx
<React.Fragment>
  <div style={{ display: 'flex', flexWrap: 'wrap' }}>
    <Icon icon="ArrowDown" />
    <Icon icon="ArrowDownTrimmed" />
    <Icon icon="ArrowUp" />
    <Icon icon="ArrowUpTrimmed" />
    <Icon icon="Asset" />
    <Icon icon="AssetTrimmed" />
    <Icon icon="ChatBubble" />
    <Icon icon="ChatBubbleTrimmed" />
    <Icon icon="CheckCircle" />
    <Icon icon="CheckCircleTrimmed" />
    <Icon icon="ChevronDown" />
    <Icon icon="ChevronDownTrimmed" />
    <Icon icon="ChevronLeft" />
    <Icon icon="ChevronLeftTrimmed" />
    <Icon icon="ChevronRight" />
    <Icon icon="ChevronRightTrimmed" />
    <Icon icon="ChevronUp" />
    <Icon icon="ChevronUpTrimmed" />
    <Icon icon="Clock" />
    <Icon icon="ClockTrimmed" />
    <Icon icon="Close" />
    <Icon icon="CloseTrimmed" />
    <Icon icon="Code" />
    <Icon icon="CodeTrimmed" />
    <Icon icon="Copy" />
    <Icon icon="CopyTrimmed" />
    <Icon icon="Download" />
    <Icon icon="DownloadTrimmed" />
    <Icon icon="Drag" />
    <Icon icon="DragTrimmed" />
    <Icon icon="Edit" />
    <Icon icon="EditTrimmed" />
    <Icon icon="EmbeddedEntryBlock" />
    <Icon icon="EmbeddedEntryBlockTrimmed" />
    <Icon icon="EmbeddedEntryInline" />
    <Icon icon="EmbeddedEntryInlineTrimmed" />
    <Icon icon="Entry" />
    <Icon icon="EntryTrimmed" />
    <Icon icon="Environment" />
    <Icon icon="EnvironmentAlias" />
    <Icon icon="ErrorCircle" />
    <Icon icon="ErrorCircleTrimmed" />
    <Icon icon="ExternalLink" />
    <Icon icon="ExternalLinkTrimmed" />
    <Icon icon="FaceHappy" />
    <Icon icon="FaceHappyTrimmed" />
    <Icon icon="Filter" />
    <Icon icon="FilterTrimmed" />
    <Icon icon="Folder" />
    <Icon icon="FolderCreate" />
    <Icon icon="FolderCreateTrimmed" />
    <Icon icon="FolderOpen" />
    <Icon icon="FolderOpenTrimmed" />
    <Icon icon="FolderTrimmed" />
    <Icon icon="FormatBold" />
    <Icon icon="FormatBoldTrimmed" />
    <Icon icon="FormatItalic" />
    <Icon icon="FormatItalicTrimmed" />
    <Icon icon="FormatUnderlined" />
    <Icon icon="FormatUnderlinedTrimmed" />
    <Icon icon="Heading" />
    <Icon icon="HeadingOne" />
    <Icon icon="HeadingOneTrimmed" />
    <Icon icon="HeadingTrimmed" />
    <Icon icon="HeadingTwo" />
    <Icon icon="HeadingTwoTrimmed" />
    <Icon icon="HelpCircle" />
    <Icon icon="HelpCircleInverted" />
    <Icon icon="HelpCircleTrimmed" />
    <Icon icon="HorizontalRule" />
    <Icon icon="HorizontalRuleTrimmed" />
    <Icon icon="InfoCircle" />
    <Icon icon="InfoCircleTrimmed" />
    <Icon icon="Link" />
    <Icon icon="LinkTrimmed" />
    <Icon icon="ListBulleted" />
    <Icon icon="ListBulletedTrimmed" />
    <Icon icon="ListNumbered" />
    <Icon icon="ListNumberedTrimmed" />
    <Icon icon="Lock" />
    <Icon icon="LockTrimmed" />
    <Icon icon="LooksOne" />
    <Icon icon="LooksOneTrimmed" />
    <Icon icon="LooksTwo" />
    <Icon icon="LooksTwoTrimmed" />
    <Icon icon="Menu" />
    <Icon icon="MenuTrimmed" />
    <Icon icon="MoreHorizontal" />
    <Icon icon="MoreHorizontalTrimmed" />
    <Icon icon="MoreVertical" />
    <Icon icon="MoreVerticalTrimmed" />
    <Icon icon="Plus" />
    <Icon icon="PlusCircle" />
    <Icon icon="PlusCircleTrimmed" />
    <Icon icon="PlusTrimmed" />
    <Icon icon="Quote" />
    <Icon icon="QuoteTrimmed" />
    <Icon icon="Receipt" />
    <Icon icon="ReceiptTrimmed" />
    <Icon icon="Search" />
    <Icon icon="SearchTrimmed" />
    <Icon icon="Settings" />
    <Icon icon="SettingsTrimmed" />
    <Icon icon="ShoppingCart" />
    <Icon icon="ShoppingCartTrimmed" />
    <Icon icon="Star" />
    <Icon icon="StarTrimmed" />
    <Icon icon="Text" />
    <Icon icon="TextTrimmed" />
    <Icon icon="ThumbDown" />
    <Icon icon="ThumbDownTrimmed" />
    <Icon icon="ThumbUp" />
    <Icon icon="ThumbUpTrimmed" />
    <Icon icon="Users" />
    <Icon icon="UsersTrimmed" />
    <Icon icon="Warning" />
    <Icon icon="WarningTrimmed" />
  </div>
</React.Fragment>
```

## Accessibility

Missing
