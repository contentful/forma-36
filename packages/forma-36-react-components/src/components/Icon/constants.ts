export const iconName = {
  ArrowDown: 'ArrowDown',
  ArrowDownTrimmed: 'ArrowDownTrimmed',
  ArrowUp: 'ArrowUp',
  ArrowUpTrimmed: 'ArrowUpTrimmed',
  Asset: 'Asset',
  AssetTrimmed: 'AssetTrimmed',
  Calendar: 'Calendar',
  ChatBubble: 'ChatBubble',
  ChatBubbleTrimmed: 'ChatBubbleTrimmed',
  CheckCircle: 'CheckCircle',
  CheckCircleTrimmed: 'CheckCircleTrimmed',
  ChevronDown: 'ChevronDown',
  ChevronDownTrimmed: 'ChevronDownTrimmed',
  ChevronLeft: 'ChevronLeft',
  ChevronLeftTrimmed: 'ChevronLeftTrimmed',
  ChevronRight: 'ChevronRight',
  ChevronRightTrimmed: 'ChevronRightTrimmed',
  ChevronUp: 'ChevronUp',
  ChevronUpTrimmed: 'ChevronUpTrimmed',
  Clock: 'Clock',
  ClockTrimmed: 'ClockTrimmed',
  Close: 'Close',
  CloseTrimmed: 'CloseTrimmed',
  Code: 'Code',
  CodeTrimmed: 'CodeTrimmed',
  Copy: 'Copy',
  CopyTrimmed: 'CopyTrimmed',
  Cycle: 'Cycle',
  CycleTrimmed: 'CycleTrimmed',
  Delete: 'Delete',
  DeleteTrimmed: 'DeleteTrimmed',
  Done: 'Done',
  DoubleArrow: 'DoubleArrow',
  Download: 'Download',
  DownloadTrimmed: 'DownloadTrimmed',
  Drag: 'Drag',
  DragTrimmed: 'DragTrimmed',
  Edit: 'Edit',
  EditTrimmed: 'EditTrimmed',
  EmbeddedEntryBlock: 'EmbeddedEntryBlock',
  EmbeddedEntryBlockTrimmed: 'EmbeddedEntryBlockTrimmed',
  EmbeddedEntryInline: 'EmbeddedEntryInline',
  EmbeddedEntryInlineTrimmed: 'EmbeddedEntryInlineTrimmed',
  Entry: 'Entry',
  EntryTrimmed: 'EntryTrimmed',
  Environment: 'Environment',
  EnvironmentAlias: 'EnvironmentAlias',
  ErrorCircle: 'ErrorCircle',
  ErrorCircleOutline: 'ErrorCircleOutline',
  ErrorCircleTrimmed: 'ErrorCircleTrimmed',
  ExternalLink: 'ExternalLink',
  ExternalLinkTrimmed: 'ExternalLinkTrimmed',
  FaceHappy: 'FaceHappy',
  FaceHappyTrimmed: 'FaceHappyTrimmed',
  Filter: 'Filter',
  FilterTrimmed: 'FilterTrimmed',
  Folder: 'Folder',
  FolderCreate: 'FolderCreate',
  FolderCreateTrimmed: 'FolderCreateTrimmed',
  FolderOpen: 'FolderOpen',
  FolderOpenTrimmed: 'FolderOpenTrimmed',
  FolderTrimmed: 'FolderTrimmed',
  FormatBold: 'FormatBold',
  FormatBoldTrimmed: 'FormatBoldTrimmed',
  FormatItalic: 'FormatItalic',
  FormatItalicTrimmed: 'FormatItalicTrimmed',
  FormatUnderlined: 'FormatUnderlined',
  FormatUnderlinedTrimmed: 'FormatUnderlinedTrimmed',
  Heading: 'Heading',
  HeadingOne: 'HeadingOne',
  HeadingOneTrimmed: 'HeadingOneTrimmed',
  HeadingTrimmed: 'HeadingTrimmed',
  HeadingTwo: 'HeadingTwo',
  HeadingTwoTrimmed: 'HeadingTwoTrimmed',
  HelpCircle: 'HelpCircle',
  HelpCircleInverted: 'HelpCircleInverted',
  HelpCircleTrimmed: 'HelpCircleTrimmed',
  HorizontalRule: 'HorizontalRule',
  HorizontalRuleTrimmed: 'HorizontalRuleTrimmed',
  InfoCircle: 'InfoCircle',
  InfoCircleTrimmed: 'InfoCircleTrimmed',
  Language: 'Language',
  Link: 'Link',
  LinkTrimmed: 'LinkTrimmed',
  ListBulleted: 'ListBulleted',
  ListBulletedTrimmed: 'ListBulletedTrimmed',
  ListNumbered: 'ListNumbered',
  ListNumberedTrimmed: 'ListNumberedTrimmed',
  Lock: 'Lock',
  LockTrimmed: 'LockTrimmed',
  Logout: 'Logout',
  LooksOne: 'LooksOne',
  LooksOneTrimmed: 'LooksOneTrimmed',
  LooksTwo: 'LooksTwo',
  LooksTwoTrimmed: 'LooksTwoTrimmed',
  Menu: 'Menu',
  MenuTrimmed: 'MenuTrimmed',
  Minus: 'Minus',
  MoreHorizontal: 'MoreHorizontal',
  MoreHorizontalTrimmed: 'MoreHorizontalTrimmed',
  MoreVertical: 'MoreVertical',
  MoreVerticalTrimmed: 'MoreVerticalTrimmed',
  Person: 'Person',
  Plus: 'Plus',
  PlusCircle: 'PlusCircle',
  PlusCircleTrimmed: 'PlusCircleTrimmed',
  PlusTrimmed: 'PlusTrimmed',
  Preview: 'Preview',
  Quote: 'Quote',
  QuoteTrimmed: 'QuoteTrimmed',
  References: 'References',
  Release: 'Release',
  ReleaseTrimmed: 'ReleaseTrimmed',
  Receipt: 'Receipt',
  ReceiptTrimmed: 'ReceiptTrimmed',
  Search: 'Search',
  SearchTrimmed: 'SearchTrimmed',
  Settings: 'Settings',
  SettingsTrimmed: 'SettingsTrimmed',
  ShoppingCart: 'ShoppingCart',
  ShoppingCartTrimmed: 'ShoppingCartTrimmed',
  Star: 'Star',
  StarTrimmed: 'StarTrimmed',
  Subscript: 'Subscript',
  SubscriptTrimmed: 'SubscriptTrimmed',
  Superscript: 'Superscript',
  SuperscriptTrimmed: 'SuperscriptTrimmed',
  Tags: 'Tags',
  Text: 'Text',
  TextTrimmed: 'TextTrimmed',
  ThumbDown: 'ThumbDown',
  ThumbDownTrimmed: 'ThumbDownTrimmed',
  ThumbUp: 'ThumbUp',
  ThumbUpTrimmed: 'ThumbUpTrimmed',
  Users: 'Users',
  UsersTrimmed: 'UsersTrimmed',
  Warning: 'Warning',
  WarningTrimmed: 'WarningTrimmed',
  Workflows: 'Workflows',
};

export const iconSizes = {
  tiny: '16px',
  small: '18px',
  medium: '24px',
  large: '32px',
};

export const iconColors = [
  'primary',
  'positive',
  'negative',
  'warning',
  'muted',
  'secondary',
  'white',
];
