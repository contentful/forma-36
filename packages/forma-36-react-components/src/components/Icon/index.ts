export { Icon } from './Icon';
export type { IconProps, IconType } from './Icon';
