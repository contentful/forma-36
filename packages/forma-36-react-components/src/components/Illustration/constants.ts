export const illustrationName = {
  Archive: 'Archive',
  Audio: 'Audio',
  Code: 'Code',
  Image: 'Image',
  Markup: 'Markup',
  Pdf: 'Pdf',
  Plaintext: 'Plaintext',
  Presentation: 'Presentation',
  Richtext: 'Richtext',
  Spreadsheet: 'Spreadsheet',
  Video: 'Video',
};
