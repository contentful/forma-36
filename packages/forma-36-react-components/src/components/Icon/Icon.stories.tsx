import React from 'react';
import { storiesOf } from '@storybook/react';
import { text, select } from '@storybook/addon-knobs';

import Icon from './Icon';
import { iconName } from './constants';

storiesOf('Components/Icon', module)
  .addParameters({
    propTypes: Icon['__docgenInfo'],
    component: Icon,
  })
  .add('Icon (default)', () => (
    <Icon
      icon={select('icon', Object.keys(iconName), Object.keys(iconName)[0])}
      size={select(
        'size',
        {
          Tiny: 'tiny',
          'Small (default)': 'small',
          Medium: 'medium',
          Large: 'large',
        },
        'small',
      )}
      color={select(
        'color',
        {
          'Primary (default)': 'primary',
          Positive: 'positive',
          Negative: 'negative',
          Warning: 'warning',
          Secondary: 'secondary',
          Muted: 'muted',
          White: 'white',
        },
        'primary',
      )}
      className={text('className', '')}
    />
  ))
  .add('Icon (all icons)', () => (
    <div
      style={{
        columns: '3',
        width: '800px',
      }}
    >
      {Object.keys(iconName)
        .filter((icon) => !icon.toLowerCase().includes('trimmed'))
        .map((icon) => (
          <div
            key={icon}
            style={{ padding: '4px', fontSize: '14px', lineHeight: '21px' }}
          >
            <Icon
              // eslint-disable-next-line @typescript-eslint/no-explicit-any
              icon={icon as any}
              style={{ marginRight: '4px', verticalAlign: 'middle' }}
            />
            <span style={{ display: 'inline-block', verticalAlign: 'middle' }}>
              {icon}
            </span>
          </div>
        ))}
    </div>
  ));
