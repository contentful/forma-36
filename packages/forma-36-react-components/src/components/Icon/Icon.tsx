import React, { Component, CSSProperties } from 'react';
import cn from 'classnames';
import { iconName } from './constants';

import {
  ArrowDown,
  ArrowDownTrimmed,
  ArrowUp,
  ArrowUpTrimmed,
  Asset,
  AssetTrimmed,
  Calendar,
  ChatBubble,
  ChatBubbleTrimmed,
  CheckCircle,
  CheckCircleTrimmed,
  ChevronDown,
  ChevronDownTrimmed,
  ChevronLeft,
  ChevronLeftTrimmed,
  ChevronRight,
  ChevronRightTrimmed,
  ChevronUp,
  ChevronUpTrimmed,
  Clock,
  ClockTrimmed,
  Close,
  CloseTrimmed,
  Code,
  CodeTrimmed,
  Copy,
  CopyTrimmed,
  Cycle,
  CycleTrimmed,
  Delete,
  DeleteTrimmed,
  Download,
  DownloadTrimmed,
  Drag,
  DragTrimmed,
  Edit,
  EditTrimmed,
  EmbeddedEntryBlock,
  EmbeddedEntryBlockTrimmed,
  EmbeddedEntryInline,
  EmbeddedEntryInlineTrimmed,
  Entry,
  EntryTrimmed,
  ErrorCircle,
  ErrorCircleTrimmed,
  ExternalLink,
  ExternalLinkTrimmed,
  FaceHappy,
  FaceHappyTrimmed,
  Filter,
  FilterTrimmed,
  Folder,
  FolderCreate,
  FolderCreateTrimmed,
  FolderOpen,
  FolderOpenTrimmed,
  FolderTrimmed,
  FormatBold,
  FormatBoldTrimmed,
  FormatItalic,
  FormatItalicTrimmed,
  FormatUnderlined,
  FormatUnderlinedTrimmed,
  Heading,
  HeadingOne,
  HeadingOneTrimmed,
  HeadingTrimmed,
  HeadingTwo,
  HeadingTwoTrimmed,
  HelpCircle,
  HelpCircleTrimmed,
  HorizontalRule,
  HorizontalRuleTrimmed,
  InfoCircle,
  InfoCircleTrimmed,
  Language,
  Link,
  LinkTrimmed,
  ListBulleted,
  ListBulletedTrimmed,
  ListNumbered,
  ListNumberedTrimmed,
  Lock,
  LockTrimmed,
  Logout,
  LooksOne,
  LooksOneTrimmed,
  LooksTwo,
  LooksTwoTrimmed,
  Menu,
  MenuTrimmed,
  MoreHorizontal,
  MoreHorizontalTrimmed,
  MoreVertical,
  MoreVerticalTrimmed,
  Person,
  Plus,
  PlusCircle,
  PlusCircleTrimmed,
  PlusTrimmed,
  Preview,
  Quote,
  QuoteTrimmed,
  Receipt,
  ReceiptTrimmed,
  References,
  Release,
  ReleaseTrimmed,
  Search,
  SearchTrimmed,
  Settings,
  SettingsTrimmed,
  ShoppingCart,
  ShoppingCartTrimmed,
  Star,
  StarTrimmed,
  Tags,
  Text,
  TextTrimmed,
  ThumbDown,
  ThumbDownTrimmed,
  ThumbUp,
  ThumbUpTrimmed,
  Users,
  UsersTrimmed,
  Warning,
  WarningTrimmed,
  Workflows,
} from './svg';

import styles from './Icon.css';

const iconComponents = {
  ArrowDown,
  ArrowDownTrimmed,
  ArrowUp,
  ArrowUpTrimmed,
  Asset,
  AssetTrimmed,
  Calendar,
  ChatBubble,
  ChatBubbleTrimmed,
  CheckCircle,
  CheckCircleTrimmed,
  ChevronDown,
  ChevronDownTrimmed,
  ChevronLeft,
  ChevronLeftTrimmed,
  ChevronRight,
  ChevronRightTrimmed,
  ChevronUp,
  ChevronUpTrimmed,
  Clock,
  ClockTrimmed,
  Close,
  CloseTrimmed,
  Code,
  CodeTrimmed,
  Copy,
  CopyTrimmed,
  Cycle,
  CycleTrimmed,
  Delete,
  DeleteTrimmed,
  Download,
  DownloadTrimmed,
  Drag,
  DragTrimmed,
  Edit,
  EditTrimmed,
  EmbeddedEntryBlock,
  EmbeddedEntryBlockTrimmed,
  EmbeddedEntryInline,
  EmbeddedEntryInlineTrimmed,
  Entry,
  EntryTrimmed,
  ErrorCircle,
  ErrorCircleTrimmed,
  ExternalLink,
  ExternalLinkTrimmed,
  FaceHappy,
  FaceHappyTrimmed,
  Filter,
  FilterTrimmed,
  Folder,
  FolderCreate,
  FolderCreateTrimmed,
  FolderOpen,
  FolderOpenTrimmed,
  FolderTrimmed,
  FormatBold,
  FormatBoldTrimmed,
  FormatItalic,
  FormatItalicTrimmed,
  FormatUnderlined,
  FormatUnderlinedTrimmed,
  Heading,
  HeadingOne,
  HeadingOneTrimmed,
  HeadingTrimmed,
  HeadingTwo,
  HeadingTwoTrimmed,
  HelpCircle,
  HelpCircleTrimmed,
  HorizontalRule,
  HorizontalRuleTrimmed,
  InfoCircle,
  InfoCircleTrimmed,
  Language,
  Link,
  LinkTrimmed,
  ListBulleted,
  ListBulletedTrimmed,
  ListNumbered,
  ListNumberedTrimmed,
  Lock,
  LockTrimmed,
  Logout,
  LooksOne,
  LooksOneTrimmed,
  LooksTwo,
  LooksTwoTrimmed,
  Menu,
  MenuTrimmed,
  MoreHorizontal,
  MoreHorizontalTrimmed,
  MoreVertical,
  MoreVerticalTrimmed,
  Person,
  Plus,
  PlusCircle,
  PlusCircleTrimmed,
  PlusTrimmed,
  Preview,
  Quote,
  QuoteTrimmed,
  Receipt,
  ReceiptTrimmed,
  References,
  Release,
  ReleaseTrimmed,
  Search,
  SearchTrimmed,
  Settings,
  SettingsTrimmed,
  ShoppingCart,
  ShoppingCartTrimmed,
  Star,
  StarTrimmed,
  Tags,
  Text,
  TextTrimmed,
  ThumbDown,
  ThumbDownTrimmed,
  ThumbUp,
  ThumbUpTrimmed,
  Users,
  UsersTrimmed,
  Warning,
  WarningTrimmed,
  Workflows,
};

export type IconType = keyof typeof iconName;
export type IconColorType =
  | 'primary'
  | 'positive'
  | 'negative'
  | 'warning'
  | 'secondary'
  | 'muted'
  | 'white';

export type IconSize = 'tiny' | 'small' | 'medium' | 'large';

export interface IconProps {
  size?: IconSize;
  color?: IconColorType;
  style?: CSSProperties;
  icon: IconType;
  className?: string;
  testId?: string;
}

const defaultProps: Partial<IconProps> = {
  testId: 'cf-ui-icon',
  size: 'small',
  color: 'primary',
};

export class Icon extends Component<IconProps> {
  static defaultProps = defaultProps;

  render() {
    const { className, icon, size, color, testId, ...otherProps } = this.props;

    const classNames = cn(
      styles.Icon,
      {
        [styles[`Icon--${size}`]]: size,
        [styles[`Icon--${color}`]]: color,
        [styles['Icon--trimmed']]: icon.toLowerCase().includes('trimmed'),
      },
      className,
    );

    const Element = iconComponents[icon];

    return (
      <Element data-test-id={testId} className={classNames} {...otherProps} />
    );
  }
}

export default Icon;
