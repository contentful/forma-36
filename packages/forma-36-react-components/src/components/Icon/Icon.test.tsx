import React from 'react';
import { shallow, mount } from 'enzyme';
import { axe } from 'jest-axe';
import Icon from './Icon';
import { iconName } from './constants';

it('renders the component', () => {
  const output = shallow(<Icon icon={'ArrowDown'} />);

  expect(output).toMatchSnapshot();
});

it('renders the component with an additional class name', () => {
  const output = shallow(
    <Icon icon={'ArrowDown'} className="my-extra-class" />,
  );

  expect(output).toMatchSnapshot();
});

it('renders a large icon', () => {
  const output = shallow(<Icon icon={'ArrowDown'} size="large" />);

  expect(output).toMatchSnapshot();
});

it('renders as a "primary" icon', () => {
  const output = shallow(
    <Icon icon={'ArrowDown'} size="large" color="primary" />,
  );

  expect(output).toMatchSnapshot();
});

it('renders as a "positive" icon', () => {
  const output = shallow(
    <Icon icon={'ArrowDown'} size="large" color="positive" />,
  );

  expect(output).toMatchSnapshot();
});

it('renders as a "negative" icon', () => {
  const output = shallow(
    <Icon icon={'ArrowDown'} size="large" color="negative" />,
  );

  expect(output).toMatchSnapshot();
});

it('renders as a "warning" icon', () => {
  const output = shallow(
    <Icon
      icon={iconName[Object.keys(iconName)[0]]}
      size="large"
      color="warning"
    />,
  );

  expect(output).toMatchSnapshot();
});

it('renders as a "secondary" icon', () => {
  const output = shallow(
    <Icon icon={'ArrowDown'} size="large" color="secondary" />,
  );

  expect(output).toMatchSnapshot();
});

it('renders as a "muted" icon', () => {
  const output = shallow(
    <Icon icon={'ArrowDown'} size="large" color="muted" />,
  );

  expect(output).toMatchSnapshot();
});

it('renders as a "white" icon', () => {
  const output = shallow(
    <Icon icon={'ArrowDown'} size="large" color="white" />,
  );

  expect(output).toMatchSnapshot();
});

Object.keys(iconName).forEach(icon => {
  it(`${icon} has no a11y issues`, async () => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const output = mount(<Icon icon={icon as any} />).html();

    const results = await axe(output);

    expect(results).toHaveNoViolations();
  });
});
