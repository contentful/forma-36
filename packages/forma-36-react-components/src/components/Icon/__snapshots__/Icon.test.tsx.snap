// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a large icon 1`] = `
<svg
  class="Icon Icon--large Icon--primary"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders as a "muted" icon 1`] = `
<svg
  class="Icon Icon--large Icon--muted"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders as a "negative" icon 1`] = `
<svg
  class="Icon Icon--large Icon--negative"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders as a "positive" icon 1`] = `
<svg
  class="Icon Icon--large Icon--positive"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders as a "primary" icon 1`] = `
<svg
  class="Icon Icon--large Icon--primary"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders as a "secondary" icon 1`] = `
<svg
  class="Icon Icon--large Icon--secondary"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders as a "warning" icon 1`] = `
<svg
  class="Icon Icon--large Icon--warning"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders as a "white" icon 1`] = `
<svg
  class="Icon Icon--large Icon--white"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders the component 1`] = `
<svg
  class="Icon Icon--small Icon--primary"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;

exports[`renders the component with an additional class name 1`] = `
<svg
  class="Icon Icon--small Icon--primary my-extra-class"
  data-test-id="cf-ui-icon"
  height="1em"
  viewBox="0 0 24 24"
  width="1em"
>
  <path
    d="M7 10l5 5 5-5z"
  />
  <path
    d="M0 0h24v24H0z"
    fill="none"
  />
</svg>
`;
